reconcilers/childset.go

```diff
@@ -0,0 +1,359 @@
+	// will run with. It's common to setup field indexes and watch resources.
+	//
+	// +optional
+	Setup func(ctx context.Context, mgr ctrl.Manager, bldr *builder.Builder) error
+
+	// DesiredChildren returns the set of desired child object for the given reconciled resource,
+	// or nil if no children should exist. Each resource returned from this method must be claimed
+	// by the OurChild method with a stable, unique identifier returned. The identifier is used to
+	// correlate desired and actual child resources.
+	//
+	// To skip reconciliation of the child resources while still reflecting an existing child's
+	// status on the reconciled resource, return OnlyReconcileChildStatus as an error.
+	DesiredChildren func(ctx context.Context, resource Type) ([]ChildType, error)
+
+	// ReflectChildrenStatusOnParent updates the reconciled resource's status with values from the
+	// child reconciliations. Select types of errors are captured, including:
+	//   - apierrs.IsAlreadyExists
+	//
+	// Most errors are returned directly, skipping this method. The set of handled error types
+	// may grow, implementations should be defensive rather than assuming the error type.
+	ReflectChildrenStatusOnParent func(ctx context.Context, parent Type, result ChildSetResult[ChildType])
+
+	// HarmonizeImmutableFields allows fields that are immutable on the current
+	// object to be copied to the desired object in order to avoid creating
+	// updates which are guaranteed to fail.
+	//
+	// +optional
+	HarmonizeImmutableFields func(current, desired ChildType)
+
+	// MergeBeforeUpdate copies desired fields on to the current object before
+	// calling update. Typically fields to copy are the Spec, Labels and
+	// Annotations.
+	MergeBeforeUpdate func(current, desired ChildType)
+
+	// ListOptions allows custom options to be use when listing potential child resources. Each
+	// resource retrieved as part of the listing is confirmed via OurChild.
+	//
+	// Defaults to filtering by the reconciled resource's namespace:
+	//     []client.ListOption{
+	//         client.InNamespace(resource.GetNamespace()),
+	//     }
+	//
+	// +optional
+	ListOptions func(ctx context.Context, resource Type) []client.ListOption
+
+	// OurChild is used when there are multiple ChildSetReconciler for the same ChildType controlled
+	// by the same reconciled resource. The function return true for child resources managed by
+	// this ChildSetReconciler. Objects returned from the DesiredChildren function should match this
+	// function, otherwise they may be orphaned.
+	//
+	// A stable identifier for the child resource must be returned when the resource is a child
+	// controlled by this reconciler. The identifier is used to correlate desired child resources
+	// with actual child resources. Non-deterministic IDs will result in the rapid deletion and
+	// creation of child resources.
+	OurChild func(resource Type, child ChildType) (bool, string)
```

Comment:
I was wondering if it's better to keep `OurChild` optional and add a new `ChildID` method for returning the ID. For most of the use cases, ownerReference is all we need to decide whether it's a child of the resource or not and the default implementation is enough. But adding id as a return type here makes it a required function to implement.

Reply:
The two are closely related in my mind. If OurChild returns true, it must also have an id. If it's not identifiable, then it's not our child.

We certainly can split them apart if the combination is more confusing than clarifying.

Reply:
added a commit splitting the two methods. wdyt

Reply:
Looks good. Thanks!